storage/provider/openstack.go

```diff
@@ -0,0 +1,399 @@
+
+type openstackVolumeSource struct {
+	openstackClient OpenstackAdapter
+}
+
+var _ storage.VolumeSource = (*openstackVolumeSource)(nil)
+
+// CreateVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) CreateVolumes(args []storage.VolumeParams) ([]storage.Volume, []storage.VolumeAttachment, error) {
+
+	vols := make([]storage.Volume, 0, len(args))
+	volAtt := make([]storage.VolumeAttachment, 0, len(args))
+
+	for _, a := range args {
+		if a.Provider != CinderProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				a.Provider,
+			)
+			continue
+		}
+
+		createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag.String())
+		if err != nil {
+			return nil, nil, err
+		}
+		vols = append(vols, createdVol)
+
+		if a.Attachment != nil {
+
+			a.Attachment.VolumeId = createdVol.VolumeId
+
+			attachments, err := s.AttachVolumes([]storage.VolumeAttachmentParams{*a.Attachment})
+			if err != nil {
+				return nil, nil, err
+			}
+
+			volAtt = append(volAtt, attachments...)
+		}
+	}
+
+	return vols, volAtt, nil
+}
+
+// DescribeVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DescribeVolumes(volIds []string) ([]storage.Volume, error) {
+
+	// In most cases, it is quicker to get all volumes and loop
+	// locally than to make several round-trips to the provider.
+	cinderVols, err := s.openstackClient.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	// Map volume ID to volume info for faster lookup.
+	idToCinderVol := make(map[string]storage.Volume)
+	for _, volInfo := range cinderVols {
+		idToCinderVol[volInfo.VolumeId] = volInfo
+	}
+
+	volumes := make([]storage.Volume, len(volIds))
+	for volIdx, volId := range volIds {
+		cinderVol, ok := idToCinderVol[volId]
+		if !ok {
+			return nil, errors.Errorf("could not find volume: %s", volId)
+		}
+		volumes[volIdx] = cinderVol
+	}
+
+	return volumes, nil
+}
+
+// DestroyVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DestroyVolumes(volIds []string) []error {
+
+	errors := make([]error, len(volIds))
+	for volIdx, volId := range volIds {
+		if err := s.openstackClient.DeleteVolume(volId); err != nil {
+			errors[volIdx] = err
+		}
+	}
+	return errors
+}
+
+// ValidateVolumeParams implements storage.VolumeSource.
+func (s *openstackVolumeSource) ValidateVolumeParams(params storage.VolumeParams) error {
+	return nil
+}
+
+// AttachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) AttachVolumes(args []storage.VolumeAttachmentParams) ([]storage.VolumeAttachment, error) {
+
+	volAttachments := make([]storage.VolumeAttachment, len(args))
+	for argIdx, attachArg := range args {
+		if attachArg.Provider != CinderProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				attachArg.Provider,
+			)
+			continue
+		}
+
+		// Check to see if the volume is already attached.
```

Comment:
I think it'd be better to just call `AttachVolume`, and then check the error result. Fewer API calls that way. Did you consider this and dismiss the option?